src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,229 @@
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.xpath;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtils;
+
+/**
+ * Generates xpath queries. Xpath queries are generated based on received
+ * {@code DetailAst} element, line number and column number.
+ *
+ * <p>
+ *     Example class
+ * </p>
+ * <pre>
+ * public class Main {
+ *
+ *     public String sayHello(String name) {
+ *         return "Hello, " + name;
+ *     }
+ * }
+ * </pre>
+ *
+ * <p>
+ *     Following expression returns list of queries. Each query is the string representing full
+ *     path to the node inside Xpath tree, whose line number is 3 and column number is 4.
+ * </p>
+ * <pre>
+ *     new XpathQueryGenerator(rootAst, 3, 4).generate();
+ * </pre>
+ *
+ * <p>
+ *     Result list
+ * </p>
+ * <ul>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS/LITERAL_PUBLIC
+ *     </li>
+ * </ul>
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathQueryGenerator {
+    /** The root ast. */
+    private final DetailAST rootAst;
+    /** The line number of the element for which the query should be generated. */
+    private final int lineNumber;
+    /** The column number of the element for which the query should be generated. */
+    private final int columnNumber;
+
+    /**
+     * Creates a new {@code XpathQueryGenerator} instance.
+     *
+     * @param rootAst root ast
+     * @param lineNumber line number of the element for which the query should be generated
+     * @param columnNumber column number of the element for which the query should be generated
+     */
+    public XpathQueryGenerator(DetailAST rootAst, int lineNumber, int columnNumber) {
+        this.rootAst = rootAst;
+        this.lineNumber = lineNumber;
+        this.columnNumber = columnNumber;
+    }
+
+    /**
+     * Returns list of xpath queries of nodes, matching line and column number.
+     * This approach uses DetailAST traversal. DetailAST means detail abstract syntax tree.
+     * @return list of xpath queries of nodes, matching line and column number
+     */
+    public List<String> generate() {
+        final List<DetailAST> matchingAstElements = getMatchingAstElements(rootAst, lineNumber,
+                columnNumber);
+        final List<String> generated = new ArrayList<>(matchingAstElements.size());
+        for (DetailAST ast : matchingAstElements) {
+            final StringBuilder xpathQueryBuilder = new StringBuilder(getXpathQuery(null, ast));
```

Comment:
WHy do we need this builder? We will create the new instances of StringBuilder each iteration in the loop. StringBuilder should be used for the cases when you concatenate strings in the loop in order to get the final string result. For example, as in getXpathQuery where you declare StringBuilder outside the loop, but here you create a string and add it to the list. String object is more lightweight than StringBuilder.